### ISSUE TITLE
MultiThread not safety

### ISSUE THREAD

**Issue by scorplot** [SDWebImage/SDWebImage#2890]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | iOS
 Platform Version        | 12.4
 SDWebImage Version      | 5.3.1
 Integration Method      | cocoapods
 Xcode Version           | Xcode 10
 Repro rate              | only once
 Repro with our demo prj | I don't try
 Demo project link       | None

### Issue Description and Steps

There are two thread access a same property which is nonatomic, this may cause memory issue.
I catched this issue by luck. I have no idea how can I reproduce.

The property is SDWebImageDownloaderOperation.dataTask.
Thread 1.  if (operation.dataTask.taskIdentifier == task.taskIdentifier) {
-[SDWebImageDownloader operationWithTask:] (in NeiTui) (SDWebImageDownloader.m:406)
-[SDWebImageDownloader URLSession:dataTask:didReceiveData:] (in NeiTui) (SDWebImageDownloader.m:436)

Thread 2.          self.dataTask = nil;
-[SDWebImageDownloaderOperation reset] (in NeiTui) (SDWebImageDownloaderOperation.m:262)
-[SDWebImageDownloaderOperation done] (in NeiTui) (SDWebImageDownloaderOperation.m:256)
__70-[SDWebImageDownloaderOperation URLSession:task:didCompleteWithError:]_block_invoke.279 (in NeiTui) (SDWebImageDownloaderOperation.m:471)

It looks like after URLSession:task:didCompleteWithError: system called URLSession:dataTask:didReceiveData:. That's maybe system issue.
I suggest [self done] do not execute in coderQueue? What do think?

**Comment by dreampiggy** [comment]
Seems the getter and setter of :

`self.dataTask` cause the tread-safe issue ?

I'll check later, if this exist, a `atomic` mark in property can solve the problem. Which is just a setter-getter lock and don't care about logic.

**Comment by dreampiggy** [comment]
Strange.

`URLSession:task:didCompleteWithError:` this is always get called on **The same delegate queue provided by URLSession, it's a global serial queue**

Why this cause thread safe issue ? The `self.dataTask` operation is always on the same queue.

**Comment by scorplot** [comment]
Hello, 
The "_**dispatch_async(self.coderQueue, ^{**_" in URLSession:task:didCompleteWithError: let "_**[self done]**_" called away from the delegate queue. 
during the **_[self done]_** the method _**[self reset]**_ should be executed and _**self.dataTask**_ has been set.
